File: tell-dont-ask-kata-kotlin/src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCaseTest.kt
package test.java.it.gabrieletondi.telldontaskkata.useCase

import main.java.it.gabrieletondi.telldontaskkata.domain.Order
import main.java.it.gabrieletondi.telldontaskkata.domain.OrderStatus
import main.java.it.gabrieletondi.telldontaskkata.useCase.OrderCannotBeShippedException
import main.java.it.gabrieletondi.telldontaskkata.useCase.OrderCannotBeShippedTwiceException
import main.java.it.gabrieletondi.telldontaskkata.useCase.OrderShipmentRequest
import main.java.it.gabrieletondi.telldontaskkata.useCase.OrderShipmentUseCase
import org.junit.Test

import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.nullValue
import org.junit.Assert.assertThat
import test.java.it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository
import test.java.it.gabrieletondi.telldontaskkata.doubles.TestShipmentService

class OrderShipmentUseCaseTest {
    private val orderRepository = TestOrderRepository()
    private val shipmentService = TestShipmentService()
    private val useCase = OrderShipmentUseCase(orderRepository, shipmentService)

    @Test
    @Throws(Exception::class)
    fun shipApprovedOrder() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.APPROVED
        orderRepository.addOrder(initialOrder)

        val request = OrderShipmentRequest()
        request.orderId = 1

        useCase.run(request)

        assertThat(orderRepository.savedOrder!!.status, `is`(OrderStatus.SHIPPED))
        assertThat(shipmentService.shippedOrder, `is`(initialOrder))
    }

    @Test(expected = OrderCannotBeShippedException::class)
    @Throws(Exception::class)
    fun createdOrdersCannotBeShipped() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.CREATED
        orderRepository.addOrder(initialOrder)

        val request = OrderShipmentRequest()
        request.orderId = 1

        useCase.run(request)

        assertThat(orderRepository.savedOrder, `is`(nullValue()))
        assertThat(shipmentService.shippedOrder, `is`(nullValue()))
    }

    @Test(expected = OrderCannotBeShippedException::class)
    @Throws(Exception::class)
    fun rejectedOrdersCannotBeShipped() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.REJECTED
        orderRepository.addOrder(initialOrder)

        val request = OrderShipmentRequest()
        request.orderId = 1

        useCase.run(request)

        assertThat(orderRepository.savedOrder, `is`(nullValue()))
        assertThat(shipmentService.shippedOrder, `is`(nullValue()))
    }

    @Test(expected = OrderCannotBeShippedTwiceException::class)
    @Throws(Exception::class)
    fun shippedOrdersCannotBeShippedAgain() {
        val initialOrder = Order()
        initialOrder.id = 1
        initialOrder.status = OrderStatus.SHIPPED
        orderRepository.addOrder(initialOrder)

        val request = OrderShipmentRequest()
        request.orderId = 1

        useCase.run(request)

        assertThat(orderRepository.savedOrder, `is`(nullValue()))
        assertThat(shipmentService.shippedOrder, `is`(nullValue()))
    }
}

File: tell-dont-ask-kata-csharp/TellDontAsk.Tests/useCase/OrderShipmentUseCaseTest.cs
using Xunit;

public class OrderShipmentUseCaseTest
{
    private readonly TestOrderRepository orderRepository = new TestOrderRepository();
    private readonly TestShipmentService shipmentService = new TestShipmentService();
    private readonly OrderShipmentUseCase useCase;

    public OrderShipmentUseCaseTest()
    {
        useCase = new OrderShipmentUseCase(orderRepository, shipmentService);
    }

    [Fact]
    public void ShipApprovedOrder()
    {
        Order initialOrder = new Order();
        initialOrder.SetId(1);
        initialOrder.SetStatus(OrderStatus.APPROVED);
        orderRepository.AddOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.SetOrderId(1);

        useCase.Run(request);

        Assert.Equal(OrderStatus.SHIPPED, orderRepository.GetSavedOrder().GetStatus());
        Assert.Equal(initialOrder, shipmentService.GetShippedOrder());
    }

    [Fact]
    public void CreatedOrdersCannotBeShipped()
    {
        Order initialOrder = new Order();
        initialOrder.SetId(1);
        initialOrder.SetStatus(OrderStatus.CREATED);
        orderRepository.AddOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.SetOrderId(1);

        Assert.Throws<OrderCannotBeShippedException>(() => useCase.Run(request));

        Assert.Null(orderRepository.GetSavedOrder());
        Assert.Null(shipmentService.GetShippedOrder());
    }

    [Fact]
    public void RejectedOrdersCannotBeShipped()
    {
        Order initialOrder = new Order();
        initialOrder.SetId(1);
        initialOrder.SetStatus(OrderStatus.REJECTED);
        orderRepository.AddOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.SetOrderId(1);

        Assert.Throws<OrderCannotBeShippedException>(() => useCase.Run(request));

        Assert.Null(orderRepository.GetSavedOrder());
        Assert.Null(shipmentService.GetShippedOrder());
    }

    [Fact]
    public void ShippedOrdersCannotBeShippedAgain()
    {
        Order initialOrder = new Order();
        initialOrder.SetId(1);
        initialOrder.SetStatus(OrderStatus.SHIPPED);
        orderRepository.AddOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.SetOrderId(1);

        Assert.Throws<OrderCannotBeShippedTwiceException>(() => useCase.Run(request));

        Assert.Null(orderRepository.GetSavedOrder());
        Assert.Null(shipmentService.GetShippedOrder());
    }
}